etna/records/models.py

```diff
@@ -1,78 +1,261 @@
-    """
-
-    iaid: str
-    title: str
-    reference_number: str
-
-    legal_status: str = ""
-    created_by: str = ""
-    description: str = ""
-    origination_date: str = ""
-    closure_status: str = ""
-    availability_delivery_condition: str = ""
-    arrangement: str = ""
-    held_by: str = ""
-    is_digitised: bool = False
-    parent: dict = field(default_factory=dict)
-    hierarchy: dict = field(default_factory=dict)
-    media_reference_id: str = ""
-    availability_delivery_surrogates: dict = field(default_factory=dict)
-    topics: dict = field(default_factory=dict)
-    next_record: dict = field(default_factory=dict)
-    previous_record: dict = field(default_factory=dict)
-    related_records: dict = field(default_factory=dict)
-    related_articles: dict = field(default_factory=dict)
-    catalogue_source: str = ""
-    repo_summary_title: str = ""
-    repo_archon_value: str = ""
-    level_code: str = ""
-    level: str = ""
-    template_reference_number: str = ""
-    template_summary_title: str = ""
-    hierarchy_level3_reference_number: str = ""
-    hierarchy_level3_summary_title: str = ""
-    related_materials: list = field(default_factory=list)
-
-    _debug_kong_result: dict = field(default_factory=dict)
+    """A 'lazy' data-interaction layer for record data retrieved from the Kong API"""
+
+    def __init__(self, raw_data: Dict[str, Any]):
+        """
+        This method recieves the raw JSON data dict recieved from
+        Kong and makes it available to the instance as `self._raw`.
+        """
+        self.score = raw_data.get("_score")
+        self._raw = raw_data.get("_source") or raw_data
+
+    @classmethod
+    def from_api_response(cls, response: dict) -> Record:
+        return cls(response)
 
     def __str__(self):
         return f"{self.title} ({self.iaid})"
 
-    def get(self, key: str, default: Optional[Any] = None):
-        return getattr(self, key, default)
+    def get(self, key: str, default: Optional[Any] = "", fail_loudly=False):
+        if "." in key:
+            return extract(self._raw, key, default, fail_loudly)
+        try:
+            return self._raw[key]
+        except KeyError as e:
+            if fail_loudly:
+                raise VariableDoesNotExist(str(e))
+            return default
 
-    @classmethod
-    def from_api_response(cls, response: dict) -> Record:
-        return cls(**transform_record_result(response))
+    @cached_property
+    def template(self) -> Dict[str, Any]:
+        return self.get("@template.details", self.get("@template.results", {}))
+
+    @cached_property
+    def iaid(self) -> str:
+        try:
+            return self.template["iaid"]
+        except KeyError:
+            pass
+        admin_id = self.get("@admin.id", fail_loudly=True)
+        if re.match(IAIDConverter.regex, admin_id):
+            return admin_id
+        raise VariableDoesNotExist(f"@admin.id '{admin_id}' is not a valid iaid.")
```

Comment:
1. `VariableDoesNotExist(f"@admin.id '{admin_id}' is not a valid iaid.")` The Attribute and message talks about different things - a Variable that does not exist and Invalid Data. I think Invalid Data must be handled separately.
2. In order to see `VariableDoesNotExist` on browser, `/catalogue/id/C1286726/` I tried commenting out `#if re.match(IAIDConverter.regex, admin_id):  #return admin_id`. But it gets swallowed by the `utils.extract` and does not display.  Alternatively I tried giving a dummy iaid in  `Record.__init__ method` (`["@admin"]["id"]="C12345"`) but `VariableDoesNotExist` does not show in the Browser

Reply:
1. The method now raises `ValueError` when the value is not a valid `iaid`
2. We're now raising an `ValueExtractionError` instead of `VariableDoesNotExist`. Also, I have explained in person that this is necessary to avoid using the value as an `iaid` when it is not suitable.